### ISSUE TITLE
Redesign/refactor Environment

### ISSUE THREAD

**Issue by Lythenas** [sp-uulm/MiniLua#23]
Environment is currently verbose to use end at the same time also prone to memory leaks.

This is both due to `std::shared_ptr` and the inheritance from `std::enable_shared_from_this`.

We should take a closer look at how the nested scopes and the global environment (`_ENV`) in lua work. It is probably possible to get rid of the `shared_ptr` completely or at least on the external interface of `Environment`.

In particular I think it is important to get rid of the constructor `Environment(const shared_ptr<Environment>& parent)` because it is very easy to cause a memory leak if somewhere in the program there is code something like the following (likely not close together or in the same function):

```cpp
// create an environment
auto env = std::make_shared<lua::rt::Environment>(nullptr);

// somewhere else
*env = env;
```

Also the shared_ptr is not even properly used in the code because most of the time functions take `const shared_ptr<Environment>&` as an argument. We could use `const Environment&` just as easy (or probably easier) and this would also prevent the `Environment` from being deconstructed (because of the `const`).

**Comment by ThomasWitte** [comment]
Making the constructor explicit would probably help avoiding this problem

**Comment by Lythenas** [comment]
Most of the points above are fixed by the new `Environment` public api. Additionally the refactoring of the environment should also fix #25. This is done by separating the global and local environments.

Also the local environment should not be accessible in the public api and (probably) needs some internals of the interpreter or at least interacts very tightly with the interpreter. So this is implemented in a separate `Env` class that is not in the public api.
